### PR TITLE
Elasticsearch ACLs

### PR DESCRIPTION
**Pull request by ivan-savciuc**

### CODE REVIEW

File: elasticsearch_acls.go [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+	bts, err := h.client.doPutRequest(path, req)
+	if err != nil {
+		return nil, err
+	}
+
+	return h.response(bts)
+}
+
+// List lists all existing Elasticsearch ACLs config
+func (h *ElasticSearchACLsHandler) List(project, service string) (*ElasticSearchACLResponse, error) {
```
Comment (mteiste) [inline elasticsearch_acls.go]: I still think this should be Get as there is only one config for each service

File: elasticsearch_acls.go [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+				}
+			}
+		}
+	}
+
+	return conf
+}
+
+// Check checks Elasticsearch permission and rules
+func (conf ElasticSearchACLConfig) Check() error {
```
Comment (mteiste) [inline elasticsearch_acls.go]: I would recommend dropping this as the API will do the checking

File: elasticsearch_acls.go (hunk 1 above)
Comment (ivan-savciuc) [reply elasticsearch_acls.go]: ok, I will change that

File: elasticsearch_acls.go (hunk 2 above)
Comment (ivan-savciuc) [reply elasticsearch_acls.go]: done

File: elasticsearch_acls.go [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+
+	return h.response(bts)
+}
+
+// Add function adds one new ACL record to already existing Elasticsearch ACL configuration,
+// it first checks if such ES user already exists. And if yes goes through the list of
+// ACL rules for such user and check if new rules that user tries to add already exists in
+// the configuration or not. We only add new records to the Elasticsearch configuration if case
+// there is a difference between actual existing configuration and what user is willing to add.
+func (h *ElasticSearchACLsHandler) Add(project, service string, acl ElasticSearchACL) error {
```
Comment (mteiste) [inline elasticsearch_acls.go]: Is this actually needed?

File: elasticsearch_acls.go [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+	}
+
+	_, err = h.Update(project, service, ElasticsearchACLRequest{
+		ElasticSearchACLConfig: acls.ElasticSearchACLConfig.Add(acl),
+	})
+	return err
+}
+
+// Delete deletes one ACL record (might contain multiple rules) from Elasticsearch ACLs list
+func (h *ElasticSearchACLsHandler) Delete(project, service string, acl ElasticSearchACL) error {
```
Comment (mteiste) [inline elasticsearch_acls.go]: same here, is this really needed?

File: elasticsearch_acls.go [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+			}
+		}
+	}
+
+	if len(acl.Rules) == 0 {
+		return conf // nothing to add to already existing ACL rules list for a user
+	}
+
+	// add to existing Elasticsearch ACL config new records
+	conf.Enabled = true
```
Comment (mteiste) [inline elasticsearch_acls.go]: It should be possible to add without enabling.

File: elasticsearch_acls.go (hunk 3 above)
Comment (ivan-savciuc) [reply elasticsearch_acls.go]: I reckon yes, imagine a situation when we have a company which is actively using Elasticsearch ACLs. Lets they created a terraform script which generates user A with some permissions. Still, some other employees from this company go to the web interface and add new user B.

 If we did not provide Add method, they would need to fire Update with entire config which based on Terraform script only has user A, and when they run Terraform apply next time user B will be deleted from ACLs list.

File: elasticsearch_acls.go (hunk 5 above)
Comment (ivan-savciuc) [reply elasticsearch_acls.go]: ok, good to know

File: elasticsearch_acls.go (hunk 3 above)
Comment (ivan-savciuc) [reply elasticsearch_acls.go]: For example, Kafka ALC API endpoint works differently from Elasticsearch ACLs, when user whats to add new Kafka ACL they send a POST request to `Kafka/ACL` and new ACL gets appended to Kafka ACL list. But Elasticsearch ACLs endpoint work differently, the user always has to provide the entire list of ACLs and these new lists will replace the old one. API interface does not offer functionality to append or add new ACL to Elasticssearch ACLlist that is why I belave these methods are required.

**Pull request merged**